File: legacy/packages/contracts/python/src/gta_ai_contracts/editing.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field

from .common import JsonValue, NonNegativeInt, PositiveInt, StudioModel, UnitScore


class SubjectTrackPoint(StudioModel):
    id: UUID
    segment_id: UUID | None = None
    frame_id: UUID | None = None
    timestamp_ms: NonNegativeInt
    focus_x: UnitScore
    focus_y: UnitScore
    confidence: UnitScore
    method: Literal["evidence_region", "visual_attention", "combined", "center_fallback"]
    source_type: str


class AdvancedEditClip(StudioModel):
    index: NonNegativeInt
    start_ms: NonNegativeInt
    end_ms: PositiveInt
    source_duration_ms: PositiveInt
    duration_ms: PositiveInt
    supporting_segment_ids: list[UUID]
    selection_score: UnitScore | None = None
    speed: float = Field(ge=1, le=2)
    speed_reason: str
    reframe_mode: Literal["dynamic_crop", "fixed_crop", "blur_background", "split_screen"]
    focus_start_x: UnitScore
    focus_end_x: UnitScore
    focus_y: UnitScore
    tracking_confidence: UnitScore
    tracking_method: Literal["evidence_region", "visual_attention", "combined", "center_fallback"]
    zoom: float = Field(ge=1, le=1.2)
    zoom_reason: str
    comparison: dict[str, JsonValue] | None = None
    concepts: list[str] = Field(default_factory=list)
    purposes: list[str] = Field(default_factory=list)
    fade_in_ms: NonNegativeInt | None = None
    fade_out_ms: NonNegativeInt | None = None


class OverlayCue(StudioModel):
    id: UUID
    cue_type: Literal["title", "step", "proof", "before_after", "result", "conclusion"]
    start_ms: NonNegativeInt
    end_ms: PositiveInt
    text: str
    secondary_text: str | None = None
    template_key: str
    supporting_claim_ids: list[UUID] = Field(default_factory=list)
    parameters: dict[str, JsonValue] = Field(default_factory=dict)


class AdvancedEditPlan(StudioModel):
    schema_version: Literal["1.0"] = "1.0"
    id: UUID
    algorithm_version: str
    project_id: UUID
    brief_id: UUID
    status: Literal["READY", "READY_WITH_FALLBACKS", "FAILED"]
    template: dict[str, JsonValue]
    safe_area: dict[str, JsonValue]
    clips: list[AdvancedEditClip]
    subject_track: list[SubjectTrackPoint]
    overlays: list[OverlayCue]
    transitions: list[dict[str, JsonValue]]
    audio_mix: dict[str, JsonValue]
    summary: dict[str, JsonValue]
    safety: dict[str, JsonValue]
    created_at: datetime

File: legacy/packages/contracts/src/editing.ts
import type { IsoDateTime, JsonObject, Uuid } from "./common.js";

export type ReframeMode = "dynamic_crop" | "fixed_crop" | "blur_background" | "split_screen";
export type TrackingMethod = "evidence_region" | "visual_attention" | "combined" | "center_fallback";
export type OverlayCueType = "title" | "step" | "proof" | "before_after" | "result" | "conclusion";

export interface SubjectTrackPoint {
  readonly id: Uuid;
  readonly segment_id: Uuid | null;
  readonly frame_id: Uuid | null;
  readonly timestamp_ms: number;
  readonly focus_x: number;
  readonly focus_y: number;
  readonly confidence: number;
  readonly method: TrackingMethod;
  readonly source_type: string;
}

export interface AdvancedEditClip {
  readonly index: number;
  readonly start_ms: number;
  readonly end_ms: number;
  readonly source_duration_ms: number;
  readonly duration_ms: number;
  readonly supporting_segment_ids: readonly Uuid[];
  readonly selection_score?: number;
  readonly speed: number;
  readonly speed_reason: string;
  readonly reframe_mode: ReframeMode;
  readonly focus_start_x: number;
  readonly focus_end_x: number;
  readonly focus_y: number;
  readonly tracking_confidence: number;
  readonly tracking_method: TrackingMethod;
  readonly zoom: number;
  readonly zoom_reason: string;
  readonly comparison?: JsonObject;
  readonly concepts: readonly string[];
  readonly purposes: readonly string[];
  readonly fade_in_ms?: number;
  readonly fade_out_ms?: number;
}

export interface OverlayCue {
  readonly id: Uuid;
  readonly cue_type: OverlayCueType;
  readonly start_ms: number;
  readonly end_ms: number;
  readonly text: string;
  readonly secondary_text: string | null;
  readonly template_key: string;
  readonly supporting_claim_ids: readonly Uuid[];
  readonly parameters: JsonObject;
}

export interface AdvancedEditPlan {
  readonly id: Uuid;
  readonly schema_version: "1.0";
  readonly algorithm_version: string;
  readonly project_id: Uuid;
  readonly brief_id: Uuid;
  readonly status: "READY" | "READY_WITH_FALLBACKS" | "FAILED";
  readonly template: JsonObject;
  readonly safe_area: JsonObject;
  readonly clips: readonly AdvancedEditClip[];
  readonly subject_track: readonly SubjectTrackPoint[];
  readonly overlays: readonly OverlayCue[];
  readonly transitions: readonly JsonObject[];
  readonly audio_mix: JsonObject;
  readonly summary: JsonObject;
  readonly safety: JsonObject;
  readonly created_at: IsoDateTime;
}
